Docs and Tests for XDS-relay integration (#5212)
solo-io/gloo: docs/content/operations/production_deployment/_index.md

## XDS Relay

{{% notice warning %}}
Today, this feature does not work in tandem with the follow non-default installation modes of Gloo Edge: rest eds, gloo mTLS mode, gloo with istio mTLS mode
{{% /notice %}}

You should consider protecting against control plane downtime (either by node failure or gloo bug) by installing gloo edge alongside the xds-relay helm chart. This helm chart installs a daemonset (configurable, daemonset by default) of xds-relay pods that serve as intermediaries between envoy and gloo (the xds server).

This serves two purposes. One, it separates the lifecycle of gloo edge from the xds cache proxies. This means a failure during helm upgrade will not mean loss of last good xds state. Second, it allows you to scale xds-relay to as many replicas as desired, since gloo is only intended for one replica today. Without xds-relay, if the single gloo replica were down, then any new envoy proxies spun up would be unable to get valid configuration.

To enable:

- install `xds-relay`:
  - `helm repo add xds-relay https://storage.googleapis.com/xds-relay-helm`

>   - `helm install xdsrelay xds-relay/xds-relay`

Reviewer: Is there any additional configuration that is exposed to the user for this install? Is this only available with the 1.9 beta?

Author: https://github.com/solo-io/gloo/pull/5212/commits/f9fd2dcb434ef36c809679d9c4fe52de6561608b

A helpful [overview of this feature is available in Envoy's documentation](https://www.envoyproxy.io/docs/envoy/latest/intro/arch_overview/upstream/outlier).
This can be configured via the `outlierDetection` field on the `Upstream` resource. See the {{< protobuf name="gloo.solo.io.Upstream" display="API reference for more detail" >}}.

## XDS Relay

{{% notice warning %}}

> Today, this feature does not work in tandem with the follow non-default installation modes of Gloo Edge: rest eds, gloo mTLS mode, gloo with istio mTLS mode

Reviewer 2: ```suggestion
This feature is not supported for the following non-default installation modes of Gloo Edge: REST Endpoint Discovery (EDS), Gloo Edge mTLS mode, Gloo Edge with Istio mTLS mode
```

Author: https://github.com/solo-io/gloo/pull/5212/commits/f26d3b115c64d749b48709ce8618e1cc2ba3a51e

{{% /notice %}}

> You should consider protecting against control plane downtime (either by node failure or gloo bug) by installing gloo edge alongside the xds-relay helm chart. This helm chart installs a daemonset (configurable, daemonset by default) of xds-relay pods that serve as intermediaries between envoy and gloo (the xds server).

Reviewer 2: ```suggestion
To protect against control plane downtime, you can install Gloo Edge alongside the `xds-relay` Helm chart. This Helm chart installs a daemonset of `xds-relay` pods that serve as intermediaries between Envoy proxies and the xDS server of Gloo Edge.
```

Author: https://github.com/solo-io/gloo/pull/5212/commits/f26d3b115c64d749b48709ce8618e1cc2ba3a51e

> This serves two purposes. One, it separates the lifecycle of gloo edge from the xds cache proxies. This means a failure during helm upgrade will not mean loss of last good xds state. Second, it allows you to scale xds-relay to as many replicas as desired, since gloo is only intended for one replica today. Without xds-relay, if the single gloo replica were down, then any new envoy proxies spun up would be unable to get valid configuration.

Reviewer 2: ```suggestion
The presence of `xds-relay` intermediary pods serve two purposes. First, it separates the lifecycle of Gloo Edge from the xDS cache proxies. For example, a failure during a Helm upgrade will not cause the loss of the last valid xDS state. Second, it allows you to scale `xds-relay` to as many replicas as needed, since Gloo Edge uses only one replica. Without `xds-relay`, a failure of the single Gloo Edge replica causes any new Envoy proxies to be created without a valid configuration.
```

Author: https://github.com/solo-io/gloo/pull/5212/commits/f26d3b115c64d749b48709ce8618e1cc2ba3a51e

To enable:

- install `xds-relay`:
  - `helm repo add xds-relay https://storage.googleapis.com/xds-relay-helm`

>   - `helm install xdsrelay xds-relay/xds-relay` (supports envoy v2 and v3 api; should work with any Gloo Edge release)

Reviewer 2: (couldnt add a suggestion here, the formatting was giving me issues)

To install the `xds-relay` Helm chart, which supports version 2 and 3 of the Envoy API:

Add the `xds-relay` Helm repository.
```shell
helm repo add xds-relay https://storage.googleapis.com/xds-relay-helm
```

Author: https://github.com/solo-io/gloo/pull/5212/commits/f26d3b115c64d749b48709ce8618e1cc2ba3a51e

> Some extra helm values you may want to customize for xds-relay (default helm values listed below):

Reviewer 2: If needed, change the default values for the `xds-relay` chart:

Author: https://github.com/solo-io/gloo/pull/5212/commits/f26d3b115c64d749b48709ce8618e1cc2ba3a51e

```yaml
deployment:
  kind: DaemonSet
  logLevel: INFO
  image:
    pullPolicy: IfNotPresent
    registry: quay.io/solo-io
    repository: xds-relay
    tag: %version% # this will match the release version of the xds-relay chart
# might want to set resources for prod deploy, e.g.:
#  resources:
#    requests:
#      cpu: 125m
#      memory: 256Mi
service:
  type: ClusterIP
  port: 9991
bootstrap:
  originServer:
    address: gloo.gloo-system.svc.cluster.local
    port: 9977
  logging:
    level: INFO
# might want to add extra, non-default identifiers
#extraLabels:
#  k: v
#extraTemplateAnnotations:
#  k: v
```

> - install gloo edge with the following helm values for each proxy (envoy) to point them towards xds-relay:

Reviewer 2: (same here sorry couldnt give suggestion formatting)

Install the Helm chart.
```
helm install xdsrelay xds-relay/xds-relay
```

Install the Gloo Edge Helm chart with the following values for each Envoy proxy to point to `xds-relay`:

Author: https://github.com/solo-io/gloo/pull/5212/commits/f26d3b115c64d749b48709ce8618e1cc2ba3a51e
